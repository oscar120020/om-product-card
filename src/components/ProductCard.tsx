import { useProduct } from "../hooks/useProduct";
import styles from "../styles/styles.module.css";
import React, { createContext } from "react";
import { InitialValues, onChageArgs, Product, ProductCardHandlers, ProductContextProps } from "../interfaces/interfaces";

export interface Props {
  product: Product;
  // children?: React.ReactElement | React.ReactElement[];
  children?: (args: ProductCardHandlers) => JSX.Element
  className?: string;
  style?: React.CSSProperties;
  onChange?: (args: onChageArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}

export const ProductContext = createContext({} as ProductContextProps)
const { Provider } = ProductContext

export const ProductCard = ({ product, children, className, style, onChange, value, initialValues }: Props) => {
  const { counter, increaseBy, maxCount, isMaxCounterReached, reset } = useProduct({onChange, product, value, initialValues});

  return (
    <Provider value={{
      counter,
      increaseBy,
      product,
      isMaxCounterReached
    }} >
      <div style={style} className={`${styles.productCard} ${className}`}>
        { children && 
          children({
            count: counter,
            increaseBy,
            isMaxCounterReached,
            product,
            reset,
            maxCount
          }) 
        }
      </div>
    </Provider>
  );
};

