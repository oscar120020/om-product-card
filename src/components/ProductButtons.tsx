import React, { useContext } from 'react';
import { ProductContext } from './ProductCard';
import styles from '../styles/styles.module.css';

export interface Props {
  className?: string;
  style?: React.CSSProperties;
}

export const ProductButtons = ({ className, style }: Props) => {
  const { counter, increaseBy, isMaxCounterReached } = useContext(
    ProductContext
  );

  return (
    <div style={style} className={` ${styles.buttonsContainer} ${className}`}>
      <button className={styles.buttonMinus} onClick={() => increaseBy(-1)}>
        -
      </button>
      <div className={styles.countLabel}> {counter} </div>
      <button
        className={`${styles.buttonAdd} ${isMaxCounterReached &&
          styles.disabled}`}
        onClick={() => increaseBy(1)}
      >
        +
      </button>
    </div>
  );
};
