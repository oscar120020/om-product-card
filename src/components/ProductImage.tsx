import React, { useContext } from "react";
import { ProductContext } from "./ProductCard";
import styles from "../styles/styles.module.css";
import noImage from "../assets/no-image.jpg";

export interface Props {
  image?: string;
  className?: string;
  style?: React.CSSProperties
}

export const ProductImage = ({ image, className, style }: Props) => {
  
    const { product } = useContext(ProductContext)
    let imageToShow: string;
  
    if(image){
      imageToShow = image
    }else if(product.image){
      imageToShow = product.image
    }else{
      imageToShow = noImage
    }
  
    return (
      <img
        className={` ${className} ${styles.productImg}`}
        style={style}
        src={imageToShow}
        alt="coffe"
      />
    );
  };