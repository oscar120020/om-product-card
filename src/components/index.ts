import { ProductCard as ProductCardComponent } from "./ProductCard";

import { ProductButtons } from "./ProductButtons";
import { ProductImage } from "./ProductImage";
import { ProductTitle } from "./ProductTitle";
import { ProductHOCProps } from "../interfaces/interfaces";

export { ProductButtons } 
export { ProductImage } 
export { ProductTitle } 

export const ProductCard: ProductHOCProps = Object.assign( ProductCardComponent, {
    Title: ProductTitle,
    Image: ProductImage,
    Buttons: ProductButtons
})


