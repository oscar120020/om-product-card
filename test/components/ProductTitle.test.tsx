
import React from 'react'
import renderer from 'react-test-renderer'
import { ProductCard, ProductTitle } from '../../src/components'
import { product1 } from '../data/products'

describe("ProductTitle", () => {
    test('debe mostrar el componente correctamente con el titulo personalizado', () => { 
        const wrapper = renderer.create(
            <ProductTitle title='custom title' />
        )
        expect(wrapper.toJSON()).toMatchSnapshot()
    })

    test('debe mostrar el componente ProductCard con el ProductTitle sin props', () => { 
        const wrapper = renderer.create(
            <ProductCard
                product={product1}
            >   
                {
                    () => (
                        <ProductTitle/>
                    )
                }
            </ProductCard>
        )
        expect(wrapper.toJSON()).toMatchSnapshot()
    })
})