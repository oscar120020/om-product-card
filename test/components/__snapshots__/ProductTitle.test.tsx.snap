// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductTitle debe mostrar el componente ProductCard con el ProductTitle sin props 1`] = `
<div
  className="productCard undefined"
>
  <span
    className="productDescription undefined "
  >
    Coffe Mug - with Image
  </span>
</div>
`;

exports[`ProductTitle debe mostrar el componente correctamente con el titulo personalizado 1`] = `
<span
  className="productDescription undefined "
>
  custom title
</span>
`;
